### PR TITLE
[Core][Dataset] adding example for large scale data ingestion

### PR DESCRIPTION
**Pull request by scv119**
This is the effort to add an example/tutorial for user interested using Ray Dataset for large scale data ingestion. The goal is to 

- add an example user could just copy/download and run.
- also give best practice to scale the example to shuffle 500GB of data

screenshot: 

_Users_chenshen_Documents_source_ray_doc__build_html_data_examples_big_data_ingestion html

### CODE REVIEW

File: doc/examples/big_data_ingestion.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+In this example, you will learn how to build, deploy and scale up a machine
+learning shuffle ingestion pipeline using
+`Ray Dataset <https://docs.ray.io/en/latest/data/dataset.html>`_ and
+`Dataset Pipelines <https://docs.ray.io/en/latest/data/dataset-pipeline.html>`_.
+
+In particular, we will show you:
+
+* How to build a shuffle ingestion pipeline that loads, shuffles and feeds data
+  into distributed trainers in few lines of code;
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:14]:
```suggestion
  into distributed trainers in a few lines of code;
```

File: doc/examples/big_data_ingestion.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+#######################################################################
+# Build shuffle ingestion pipeline
+# ----------------------------------
+#
+# A typical machine learning ingestion pipeline consists of the following 4
+# steps:
+#
+# 1. Load the training data from external storage;
+# 2. Iterate the data for multiple epochs;
+# 3. In each epoch, applying global shuffle to normalize the data;
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:45]:
```suggestion
# 3. In each epoch, applying global shuffle to decorrelate the data;
```

File: doc/examples/big_data_ingestion.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+.. image:: ../data/dataset-repeat-2.svg
+    :align: center
+
+"""
+
+###############################################################################
+# Python Setup
+# ------------
+#
+# First, we'll import all the libraries we'll be using. This step also helps us
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:27]:
```suggestion
# First, we'll import all of the libraries we'll be using. This step also helps us
```

File: doc/examples/big_data_ingestion.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        .random_shuffle()\
+        .split(num_shards, equal=True)
+
+
+############################################################################
+# We’ve now defined a ``create_shuffle_pipeline`` function that creates an
+# ingestion pipeline.
+# It reads ``training_data_dir``, iterates for ``num_epochs`` times,
+# where in each epoch it
+# shuffles and splits the result into ``num_shards``.
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:66]:
```suggestion
# shuffles and splits the training data into ``num_shards``.
```

File: doc/examples/big_data_ingestion.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+# -----------------------------------
+# Let’s also implement a ``TrainingWorker`` which consumes the shuffled data
+# from each shard.
+#
+# For simplicity, we will define a
+# `Ray Actor <https://docs.ray.io/en/latest/actors.html>`_ that emulates
+# training workers. Specifically,
+#
+# 1. It takes one shard of the shuffle pipeline for training;
+# 2. It iterates the shard to get per epoch training dataset;
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:79]:
```suggestion
# 2. It iterates over the shard to get a training dataset per epoch;
```

File: doc/examples/big_data_ingestion.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+#######################################################################
+# Build shuffle ingestion pipeline
+# ----------------------------------
+#
+# A typical machine learning ingestion pipeline consists of the following 4
+# steps:
+#
+# 1. Load the training data from external storage;
+# 2. Iterate the data for multiple epochs;
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:44]:
```suggestion
# 2. Iterate over the data for multiple epochs;
```

File: doc/examples/big_data_ingestion.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+class TrainingWorker:
+    def __init__(self, rank: int, shard: DatasetPipeline):
+        self.rank = rank
+        self.shard = shard
+
+    def train(self):
+        for epoch, training_dataset in enumerate(self.shard):
+            # Following code emulates epoch based SGD training.
+            # TODO: replace the code for real training.
+            for i, batch in enumerate(training_dataset.iter_batches()):
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:93]: Might want to give a concrete batch size here since `iter_batches()` will return full blocks by default, which will most likely be skewed.
```suggestion
            for i, batch in enumerate(training_dataset.iter_batches(batch_size=25000)):
```

File: doc/examples/big_data_ingestion.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                print(f"Training... epoch: {epoch}, batch: {i}")
+
+
+###########################################################################
+# Let's run it
+# -----------------------------
+# Now let’s run the data pipeline end-to-end:
+# We read parquet files from the example s3 bucket,
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:101]:
```suggestion
# We read Parquet files from the example s3 bucket,
```

File: doc/examples/big_data_ingestion.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+workers = [TrainingWorker(rank, shard) for rank, shard in enumerate(splits)]
+
+# Let's run the e2e pipeline
+ray.get([worker.train.remote() for worker in workers])
+
+#################################################################################
+# Scale the shuffle ingestion pipeline
+# --------------------------------------------------------
+#
+# Scale the shuffle ingestion pipeline is simple. With Ray, we can linearly
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:124]:
```suggestion
# Scaling the shuffle ingestion pipeline is simple. With Ray, we can linearly
```

File: doc/examples/big_data_ingestion.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+# scale the pipeline from ingesting 100MB of data to 500GB of data simply by adding
+# more machines. To achieve that, we will run the same data pipeline we
+# just defined on a 74 nodes Ray cluster.
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:127]:
```suggestion
# just defined on a 74 node Ray cluster.
```

File: doc/examples/big_data_ingestion.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+# Let’s see how we implement such pipeline using Ray Dataset:
+
+
+def create_shuffle_pipeline(training_data_dir: str, num_epochs: int,
+                            num_shards: int) -> List[DatasetPipeline]:
+
+    return ray.data.read_parquet(training_data_dir)\
+        .repeat(num_epochs)\
+        .random_shuffle()\
+        .split(num_shards, equal=True)
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:58]: Nit: I think it's more readable this way
```suggestion
    return ray.data.read_parquet(training_data_dir) \
        .repeat(num_epochs) \
        .random_shuffle() \
        .split(num_shards, equal=True)
```

File: doc/examples/big_data_ingestion.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+
+@ray.remote
+class TrainingWorker:
+    def __init__(self, rank: int, shard: DatasetPipeline):
+        self.rank = rank
+        self.shard = shard
+
+    def train(self):
+        for epoch, training_dataset in enumerate(self.shard):
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:90]:
```suggestion
        for epoch, training_dataset in enumerate(self.shard.iter_datasets()):
```

File: doc/examples/big_data_ingestion.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
+# flake8: noqa: E501
+"""
+Large-scale shuffle ingestion pipeline
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:3]: I'm wondering if this should be something like "Large-scale machine learning ingestion pipeline" or "Large-scale distributed training ingestion pipeline", since that's what this pipeline is truly providing (shuffling is just one feature of the ingestion pipeline) and this might do a better job of grabbing the attention of a user that's looking for such a pipeline, since "shuffle ingestion pipeline" might not connote ML model training for some of our users. If we're wanting to explicitly call out the shuffle since it's a big feature, then we could do something like "Large-scale per-epoch shuffling ingestion pipeline for distributed training". cc @ericl for thoughts here

**Comment by scv119** [comment]
also, any idea what we should cover for 500GB shuffle. Maybe we should mention the _spread parameters.

File: doc/examples/big_data_ingestion.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+#     $ pip install ray boto3
+#     $ ray up big_data_ingestion.yaml
+#
+# **Step 2:**  Run ``big_data_ingestion.py`` on the cluster.
+#
+# .. code-block:: bash
+#
+#     $ ray rsync_up big_data_ingestion.yaml "./big_data_ingestion.py" \
+#       "~/big_data_ingestion.py"
+#     $ ray exec big_data_ingestion.yaml "~/big_data_ingestion.py"
```
Comment (clarkzinzow) [inline doc/examples/big_data_ingestion.py:146]: Does [`ray submit`](https://docs.ray.io/en/master/package-ref.html#ray-submit) not work here?
```suggestion
#     $ ray submit big_data_ingestion.yaml ./big_data_ingestion.py
```

File: doc/examples/big_data_ingestion.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
+# flake8: noqa: E501
```
Comment (ericl) [inline doc/examples/big_data_ingestion.py:1]: Could we test this in a small-scale mode in unit tests too?

File: doc/examples/big_data_ingestion.py [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+    def __init__(self, rank: int, shard: DatasetPipeline):
+        self.rank = rank
+        self.shard = shard
+
+    def train(self):
+        for epoch, training_dataset in enumerate(self.shard):
+            # Following code emulates epoch based SGD training.
+            # TODO: replace the code for real training.
+            for i, batch in enumerate(training_dataset.iter_batches()):
+                print(f"Training... epoch: {epoch}, batch: {i}")
```
Comment (ericl) [inline doc/examples/big_data_ingestion.py:94]: Can we have the example be end to end with ML Training (i.e., use Horovod or SGDv2 integration). cc @richardliaw @amogkam

File: doc/examples/big_data_ingestion.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+# Now let’s run the data pipeline end-to-end:
+# We read parquet files from the example s3 bucket,
+# create the shuffle pipeline,
+# and use training workers to consume the pipeline.
+
+NUM_WORKERS = 4
+NUM_EPOCHS = 5
+
+TRAINING_DATA_100MB = "s3://ingest-100MB-example/"
+TRAINING_DATA_500GB = "s3://ingest-500GB-example/"
```
Comment (ericl) [inline doc/examples/big_data_ingestion.py:109]: Hmm are we intending to host this ourselves? It may be better to have instructions on how to create synthetic data / use in-memory synthetic datasource by default.

File: doc/examples/big_data_ingestion.py (hunk 16 above)
Comment (amogkam) [reply doc/examples/big_data_ingestion.py:94]: Yes we should definitely show SGD v2 here once the integration with Datasets is complete.

File: doc/examples/big_data_ingestion.py (hunk 17 above)
Comment (scv119) [reply doc/examples/big_data_ingestion.py:109]: in-memory synthetic datasource sounds like a good idea.

File: doc/examples/big_data_ingestion.py [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+# generated data.
+
+
+def create_large_shuffle_pipeline(data_size_bytes: int, num_epochs: int,
+                                  num_shards: int) -> List[DatasetPipeline]:
+    # TODO: reading data from distributed filesystem (such as S3 or HDFS).
+    return ray.data.read_datasource(
+            RangeDatasource(), block_format='arrow',
+            n=data_size_bytes // 8) \
+        .repeat(num_epochs) \
```
Comment (scv119) [inline doc/examples/big_data_ingestion.py:188]: @ericl @clarkzinzow should we include the temporary hack here?

Comment (clarkzinzow) [reply doc/examples/big_data_ingestion.py:188]: I'd say include it for now, even though it's an unseemly private API, since it's how we currently achieve satisfiable performance for larger datasets, which is what we're pitching here. We can open a PR to remove it once we've removed the private API, after we've solved the underlying scheduling problem.

Comment (clarkzinzow) [reply doc/examples/big_data_ingestion.py:188]: @ericl Do you concur here? It _is_ pretty ugly!

Comment (ericl) [reply doc/examples/big_data_ingestion.py:188]: We should include it for the sake of reproducibility (but please add a XXX comment that it should be removed).

File: doc/examples/big_data_ingestion.py [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+# RangeDatasource directly. In this way we don't need to set up S3 for storing
+# generated data.
+
+
+def create_large_shuffle_pipeline(data_size_bytes: int, num_epochs: int,
+                                  num_shards: int) -> List[DatasetPipeline]:
+    # TODO: reading data from distributed filesystem (such as S3 or HDFS).
+    return ray.data.read_datasource(
+            RangeDatasource(), block_format='arrow',
+            n=data_size_bytes // 8) \
```
Comment (ericl) [inline doc/examples/big_data_ingestion.py:187]: I think you can just `ray.data.range_arrow(n)` instead. But maybe it's better to return fake data of the same schema as the S3 data? Otherwise, it would be unrealistic / can't send to horovod.

**Comment by scv119** [comment]
@amogkam @matthewdeng are you using python3.8 + pickle5? uninstall pickle5 seems solved the problem.

**Comment by matthewdeng** [comment]
> @amogkam @matthewdeng are you using python3.8 + pickle5? uninstall pickle5 seems solved the problem.

Yeah I tried uninstalling `pickle5` as well as using different versions (`0.0.10` and `0.0.11`) but the problem still persists. 

This error only happens when I use my locally installed version of Ray (`pip install -e python`), but not when I use a nightly wheel.

**Review by ericl — COMMENTED** [review]
This looks really good. A couple last requests:
1. Can you provide the expected output / performance numbers users can see after running the example at the end? This will be helpful for users as a guidepost to see if they're seeing expected results.
2. Can you move the page from the generic examples dir under the Data/ TOC? It can be called "Example: Large-scale ML Ingest" and be after the DatasetPipeline page.

**Review by ericl — APPROVED** [review]
